### ISSUE TITLE
Architect multipart forms with file uploads result in corrupted files

### ISSUE THREAD

**Issue by ericallam** [remix-run/remix#2476]
### What version of Remix are you using?

1.3.3

### Steps to Reproduce

 - Create a new remix project with `npx create-remix`
 - Select "Just the basics"
 - Select "Architect (AWS Lambda)"
 - Select "Typescript"
 - Finish creating the project
 - Replace the contents of `app/routes/index.tsx` with the following code:

```typescript
import {
  ActionFunction,
  Form,
  unstable_createFileUploadHandler,
  unstable_parseMultipartFormData,
} from "remix";

export const action: ActionFunction = async ({ request }) => {
  const uploadHandler = unstable_createFileUploadHandler({
    maxFileSize: 10_000_000,
    directory: "public/uploads",
    file: ({ filename }) => filename,
  });

  const formData = await unstable_parseMultipartFormData(
    request,
    uploadHandler
  );

  const file = formData.get("file-upload") as File;

  if (file) {
    console.log(`File uploaded to server/public/uploads/${file.name}`);
  } else {
    console.log("No file uploaded");
  }

  return {};
};

export default function Index() {
  return (
    <Form reloadDocument method="post" encType="multipart/form-data">
      <div>
        <label htmlFor="file-upload">
          <span>Upload a file</span>
          <input
            id="file-upload"
            name="file-upload"
            type="file"
          />
        </label>

        <button type="submit">Upload image</button>
      </div>
    </Form>
  );
}
```

- Run the server locally with `npm run dev`
- Open `http://localhost:3333` in your web browser and choose an image file to upload
- Submit the form
- View the file in `<root>/server/public/uploads`

### Expected Behavior

The file should be uploaded uncorrupted

### Actual Behavior

The file that is uploaded isn't the same file as the one selected, has additional data that is "corrupting" the file

**Comment by hicksy** [comment]
I'm experiencing the same issue. 

I believe the problem is in `@remix-run/architect/server.js` as I don't think `event.body` is being parsed correctly for `multipart/form-data`, so the resulting buffer isn't the correct length

To test i uploaded a **109kb** file, the results are:

### `unstable_createFileUploadHandler`:
The saved file is corrupted and 196kb in size

###  `unstable_createMemoryUploadHandler`:

```
console.log(Buffer.byteLength(await formData.get("file").arrayBuffer()));
```

length is incorrect at 196286 bytes

### custom uploadHandler:

```
const uploadHandler = async ({
  name,
  filename,
  mimetype,
  encoding,
  stream,
}) => {
  if (name !== "file") {
    stream.resume();
    return;
  }

  const stream2buffer = await new Promise((resolve, reject) => {
    const _buf = [];

    stream.on("data", (chunk) => _buf.push(chunk));
    stream.on("end", () => resolve(Buffer.concat(_buf)));
    stream.on("error", (err) => reject(err));
  });

  console.log(Buffer.byteLength(stream2buffer));

  const key = "test";

  return JSON.stringify({ filename, key });
};
```

buffer is incorrect at 196286 bytes 


### Test mod of  `@remix-run/architect/server.js`

Adding this block into line 48 of server.js, I believe confirms that the body is parsed correctly when the content-type is `multipart/form-data`

```
if (event.headers["content-type"]?.includes("multipart/form-data")) {
  let parser = await require("lambda-multipart-parser");
  bodyWithFiles = await parser.parse(event);
  console.log(Buffer.byteLength(bodyWithFiles.files[0].content));
}
```

byte length is 108911 which is correct

**Comment by hicksy** [comment]
Here's the full server.js file I've used for testing this

```
/**
 * @remix-run/architect v1.3.3
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var node = require('@remix-run/node');
var binaryTypes = require('./binary-types.js');

/**
 * A function that returns the value to use as `context` in route `loader` and
 * `action` functions.
 *
 * You can think of this as an escape hatch that allows you to pass
 * environment/platform-specific values through to your loader/action.
 */

/**
 * Returns a request handler for Architect that serves the response using
 * Remix.
 */
function createRequestHandler({
  build,
  getLoadContext,
  mode = process.env.NODE_ENV
}) {
  let handleRequest = node.createRequestHandler(build, mode);
  return async (event, _context) => {
    let abortController = new node.AbortController();
    let request = await createRemixRequest(event, abortController);
    let loadContext = typeof getLoadContext === "function" ? getLoadContext(event) : undefined;
    let response = await handleRequest(request, loadContext);
    return sendRemixResponse(response, abortController);
  };
}
async function createRemixRequest(event, abortController) {
  let host = event.headers["x-forwarded-host"] || event.headers.host;
  let search = event.rawQueryString.length ? `?${event.rawQueryString}` : "";
  let url = new URL(event.rawPath + search, `https://${host}`);
  if(event.headers['content-type']?.includes('multipart/form-data')) {
    var parser = require('lambda-multipart-parser');
    let bodyWithFiles = await parser.parse(event);
    console.log(Buffer.byteLength(bodyWithFiles.files[0].content))
  }
  return new node.Request(url.href, {
    method: event.requestContext.http.method,
    headers: createRemixHeaders(event.headers, event.cookies),
    body: event.body && event.isBase64Encoded ? Buffer.from(event.body, "base64").toString() : event.body,
    abortController,
    signal: abortController === null || abortController === void 0 ? void 0 : abortController.signal
  });
}
function createRemixHeaders(requestHeaders, requestCookies) {
  let headers = new node.Headers();

  for (let [header, value] of Object.entries(requestHeaders)) {
    if (value) {
      headers.append(header, value);
    }
  }

  if (requestCookies) {
    headers.append("Cookie", requestCookies.join("; "));
  }

  return headers;
}
async function sendRemixResponse(response, abortController) {
  let cookies = []; // Arc/AWS API Gateway will send back set-cookies outside of response headers.

  for (let [key, values] of Object.entries(response.headers.raw())) {
    if (key.toLowerCase() === "set-cookie") {
      for (let value of values) {
        cookies.push(value);
      }
    }
  }

  if (cookies.length) {
    response.headers.delete("set-cookie");
  }

  if (abortController.signal.aborted) {
    response.headers.set("Connection", "close");
  }

  let contentType = response.headers.get("content-type");
  let isBinary = binaryTypes.isBinaryType(contentType);
  let body;
  let isBase64Encoded = false;

  if (isBinary) {
    let blob = await response.arrayBuffer();
    body = Buffer.from(blob).toString("base64");
    isBase64Encoded = true;
  } else {
    body = await response.text();
  }

  return {
    statusCode: response.status,
    headers: Object.fromEntries(response.headers),
    cookies,
    body,
    isBase64Encoded
  };
}

exports.createRemixHeaders = createRemixHeaders;
exports.createRemixRequest = createRemixRequest;
exports.createRequestHandler = createRequestHandler;
exports.sendRemixResponse = sendRemixResponse;

```

**Comment by hicksy** [comment]
I've been digging into this more, I _think_ the use of `toString()` on the buffer in two locations is causing the binary data to be skewered. 

I've modified `server/node_modules/@remix-run/architect/server.js` & `server/node_modules/@remix-run/node/parseMultipartFormData.js` and I now have an uncorrupted file 😄 

Putting together a PR so they can review and decide if they're happy this is a fix / doesn't affect anything else. 

In the meantime you could manually patch the files yourself to test:

```shell
 server/node_modules/@remix-run/architect/server.js
 ```
```js
/**
 * @remix-run/architect v1.3.3
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var node = require('@remix-run/node');
var binaryTypes = require('./binary-types.js');

/**
* A function that returns the value to use as `context` in route `loader` and
* `action` functions.
*
* You can think of this as an escape hatch that allows you to pass
* environment/platform-specific values through to your loader/action.
*/

/**
* Returns a request handler for Architect that serves the response using
* Remix.
*/
function createRequestHandler({
  build,
  getLoadContext,
  mode = process.env.NODE_ENV
}) {
  let handleRequest = node.createRequestHandler(build, mode);
  return async (event, _context) => {
    let abortController = new node.AbortController();
    let request = await createRemixRequest(event, abortController);
    let loadContext = typeof getLoadContext === "function" ? getLoadContext(event) : undefined;
    let response = await handleRequest(request, loadContext);
    return sendRemixResponse(response, abortController);
  };
}
function createRemixRequest(event, abortController) {
  let host = event.headers["x-forwarded-host"] || event.headers.host;
  let search = event.rawQueryString.length ? `?${event.rawQueryString}` : "";
  let url = new URL(event.rawPath + search, `https://${host}`);
  let isFormData = false;

  if (event.headers["content-type"] && event.headers["content-type"].includes("multipart/form-data")) {
    isFormData = true;
  }

  return new node.Request(url.href, {
    method: event.requestContext.http.method,
    headers: createRemixHeaders(event.headers, event.cookies),
    body: event.body && event.isBase64Encoded ? !isFormData ? Buffer.from(event.body, "base64").toString() : Buffer.from(event.body, "base64") : event.body,
    abortController,
    signal: abortController === null || abortController === void 0 ? void 0 : abortController.signal
  });
}
function createRemixHeaders(requestHeaders, requestCookies) {
  let headers = new node.Headers();

  for (let [header, value] of Object.entries(requestHeaders)) {
    if (value) {
      headers.append(header, value);
    }
  }

  if (requestCookies) {
    headers.append("Cookie", requestCookies.join("; "));
  }

  return headers;
}
async function sendRemixResponse(response, abortController) {
  let cookies = []; // Arc/AWS API Gateway will send back set-cookies outside of response headers.

  for (let [key, values] of Object.entries(response.headers.raw())) {
    if (key.toLowerCase() === "set-cookie") {
      for (let value of values) {
        cookies.push(value);
      }
    }
  }

  if (cookies.length) {
    response.headers.delete("set-cookie");
  }

  if (abortController.signal.aborted) {
    response.headers.set("Connection", "close");
  }

  let contentType = response.headers.get("content-type");
  let isBinary = binaryTypes.isBinaryType(contentType);
  let body;
  let isBase64Encoded = false;

  if (isBinary) {
    let blob = await response.arrayBuffer();
    body = Buffer.from(blob).toString("base64");
    isBase64Encoded = true;
  } else {
    body = await response.text();
  }

  return {
    statusCode: response.status,
    headers: Object.fromEntries(response.headers),
    cookies,
    body,
    isBase64Encoded
  };
}

exports.createRemixHeaders = createRemixHeaders;
exports.createRemixRequest = createRemixRequest;
exports.createRequestHandler = createRequestHandler;
exports.sendRemixResponse = sendRemixResponse;
```

```shell
 server/node_modules/@remix-run/node/parseMultipartFormData.js
 ```

```js
/**
 * @remix-run/node v1.3.3
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var stream = require('stream');
var Busboy = require('busboy');
var formData = require('./formData.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var Busboy__default = /*#__PURE__*/_interopDefaultLegacy(Busboy);

/**
 * Allows you to handle multipart forms (file uploads) for your app.
 *
 * @see https://remix.run/api/remix#parsemultipartformdata-node
 */

function parseMultipartFormData(request, uploadHandler) {
  return request.formData(uploadHandler);
}
async function internalParseFormData(contentType, body, abortController, uploadHandler) {
  let formData$1 = new formData.FormData();
  let fileWorkQueue = [];
  let stream$1;

  if (typeof body === "string" || Buffer.isBuffer(body)) {
    stream$1 = stream.Readable.from(body);
  } else {
    stream$1 = body;
  }

  await new Promise(async (resolve, reject) => {
    let busboy = new Busboy__default["default"]({
      highWaterMark: 2 * 1024 * 1024,
      headers: {
        "content-type": contentType
      }
    });
    let aborted = false;

    function abort(error) {
      if (aborted) return;
      aborted = true;
      stream$1.unpipe();
      stream$1.removeAllListeners();
      busboy.removeAllListeners();
      abortController === null || abortController === void 0 ? void 0 : abortController.abort();
      reject(error || new Error("failed to parse form data"));
    }

    busboy.on("field", (name, value) => {
      formData$1.append(name, value);
    });
    busboy.on("file", (name, filestream, filename, encoding, mimetype) => {
      if (uploadHandler) {
        fileWorkQueue.push((async () => {
          try {
            let value = await uploadHandler({
              name,
              stream: filestream,
              filename,
              encoding,
              mimetype
            });

            if (typeof value !== "undefined") {
              formData$1.append(name, value);
            }
          } catch (error) {
            // Emit error to busboy to bail early if possible
            busboy.emit("error", error); // It's possible that the handler is doing stuff and fails
            // *after* busboy has finished. Rethrow the error for surfacing
            // in the Promise.all(fileWorkQueue) below.

            throw error;
          } finally {
            filestream.resume();
          }
        })());
      } else {
        filestream.resume();
      }

      if (!uploadHandler) {
        console.warn(`Tried to parse multipart file upload for field "${name}" but no uploadHandler was provided.` + " Read more here: https://remix.run/api/remix#parseMultipartFormData-node");
      }
    });
    stream$1.on("error", abort);
    stream$1.on("aborted", abort);
    busboy.on("error", abort);
    busboy.on("finish", resolve);
    stream$1.pipe(busboy);
  });
  await Promise.all(fileWorkQueue);
  return formData$1;
}

exports.internalParseFormData = internalParseFormData;
exports.parseMultipartFormData = parseMultipartFormData;
```

**Comment by ericallam** [comment]
Thanks @hicksy, these patches have fixed the issue for me as well. I used [patch-package](https://www.npmjs.com/package/patch-package) with the following patches using your code from above:

[@remix-run-issue-2476-patches.zip](https://github.com/remix-run/remix/files/8362945/%40remix-run-issue-2476-patches.zip)

**Comment by hicksy** [comment]
Perfect, thanks @ericallam

**Comment by mcansh** [comment]
hey folks, this is resolved in 1.3.5 by #2495 - feel free to re-open if you're still running into issues :)
